### ISSUE TITLE
`:print` doesn't output to the notebook

### ISSUE THREAD

**Issue by robbles** [gophergala2016/gophernotes#5]
When running the `:print` command, it outputs in the terminal where the notebook server is running instead of the browser. Maybe it defaults to stderr instead of stdout?

**Comment by dwhitena** [comment]
Yes, you are right.  It would output to wherever the jupyter server is running, so you could see it in the terminal or in the docker logs.  I will work on a fix.

**Comment by dwhitena** [comment]
@robbles The source of the issue is in `commands.go` where the print is defined:

```
func actionPrint(s *Session, _ string) error {
	source, err := s.source(true)
	if err != nil {
		return err
	}

	fmt.Println(source)

	return nil
}
```

Need to redirect this to the notebook.

**Comment by JoshCheek** [comment]
TL;DR
---------

Probably the [`Eval` function](https://github.com/gophergala2016/gophernotes/blob/066c768b6dd38daf00549fb0b0c531bc18af8bb8/replpkg/repl.go#L300) in needs to be updated to allow the commands to return meta information like "now run this thing" or whatever.


More In Depth
-------------------

I looked at it a bit today, modified `actionPrint` to look like this:

```go
func actionPrint(s *Session, _ string) error {
	source, err := s.source(true)

	if err == nil {
		fmt.Println(source)
		toEval := fmt.Sprintf("fmt.Println(%q)", source)
		fmt.Println(toEval)
		_, err, _ = s.Eval(toEval) // str, err, buffer
	}

	return err
}
```

The `%q` is important in the sprintf, it enables escaping. However the output of the `s.Eval(toEval)` is never sent back across to the notebook.

The command is, itself, called from [`Eval`](https://github.com/gophergala2016/gophernotes/blob/066c768b6dd38daf00549fb0b0c531bc18af8bb8/replpkg/repl.go#L315), so in this case, its sort of recursively calling eval. When the `actionPrint` returns, it sets a boolean flag to say the input was a command ([here](https://github.com/gophergala2016/gophernotes/blob/066c768b6dd38daf00549fb0b0c531bc18af8bb8/replpkg/repl.go#L322)). The boolean flag causes it to exit the method, returning that there is nothing to run. I was able to remove the extra variable like this:

```diff
diff --git a/replpkg/repl.go b/replpkg/repl.go
index abee93f..76fd668 100644
--- a/replpkg/repl.go
+++ b/replpkg/repl.go
@@ -303,7 +303,6 @@ func (s *Session) Eval(in string) (string, error, bytes.Buffer) {
        s.clearQuickFix()
        s.storeMainBody()

-       var commandRan bool
        for _, command := range commands {
                arg := strings.TrimPrefix(in, ":"+command.name)
                if arg == in {
@@ -312,6 +311,7 @@ func (s *Session) Eval(in string) (string, error, bytes.Buffer) {

                if arg == "" || strings.HasPrefix(arg, " ") {
                        arg = strings.TrimSpace(arg)
+
                        err := command.action(s, arg)
                        if err != nil {
                                if err == ErrQuit {
@@ -319,15 +319,11 @@ func (s *Session) Eval(in string) (string, error, bytes.Buffer) {
                                }
                                errorf("%s: %s", command.name, err)
                        }
-                       commandRan = true
-                       break
-               }
-       }

-       if commandRan {
                        s.doQuickFix()
                        return "", nil, bytes.Buffer{}
                }
+       }
```

Because we always return `return "", nil, bytes.Buffer{}` in that situation, we won't ever update the notebook with the result. The [return value](https://github.com/gophergala2016/gophernotes/blob/066c768b6dd38daf00549fb0b0c531bc18af8bb8/replpkg/repl.go#L395) of this Eval function is what updates the result, in the [`HandleExecuteRequest`](https://github.com/gophergala2016/gophernotes/blob/066c768b6dd38daf00549fb0b0c531bc18af8bb8/execution.go#L57-L72) method.

**Comment by robbles** [comment]
@JoshCheek would it be easier to simply intercept stdout while running Eval, collect the result in a buffer, and send it to the notebook to display at the end?

**Comment by JoshCheek** [comment]
Possibly, but I somewhat expect other use cases to arise, where returning meta level instructions from the command will be valuable. If it's trivially easy to do, then probably go for it, but it looked like the sort of thing where, if I were more familiar with the language, I could refactor it to handle such things without too much difficult, which seems likely to have greater flexibility and elegance.

**Comment by dwhitena** [comment]
Hey guys, I'm wondering now if we should/could just bypass any Eval and just send the source string (that we retrieve from the session) in a new ZeroMQ message back to Jupyter.  Might be a quick fix.
